File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int k) {
        const int inf = 1e7;
        vector<pair<int, int>> g[n];
        for (auto &f : flights) g[f[0]].emplace_back(f[1], f[2]);
        vector<vector<int>> dp(k + 2, vector<int>(n, inf));
        dp[0][src] = 0;
        for (int edge = 1; edge <= k + 1; ++edge) {
            for (int node = 0; node < n; ++node) {
                if (dp[edge - 1][node] != inf) {
                    for (auto &next : g[node]) {
                        dp[edge][next.first] = min(dp[edge][next.first], dp[edge - 1][node] + next.second);
                    }
                }
            }
        }
        int ans = inf;
        for (int edge = 0; edge < k + 2; ++edge) {
            ans = min(ans, dp[edge][dst]);
        }
        return (ans == inf) ? -1 : ans;
    }
};

File: 0787-cheapest-flights-within-k-stops/0787-cheapest-flights-within-k-stops.java
class Pair {
    int x, y;
    Pair(int x, int y) {
        this.x = x;
        this.y = y;
    }
    int getKey() {return x;}
    int getValue() {return y;}
}

class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        ArrayList<ArrayList<Pair>> graph = new ArrayList<>(n);
        for (int i = 0; i < n; ++i) graph.add(new ArrayList<Pair>());
        for (int i = 0; i < flights.length; ++i) {
            int u = flights[i][0], v = flights[i][1], w = flights[i][2];
            graph.get(u).add(new Pair(v, w));
        }
        int[][] dp = new int[n][k + 2];
        for (int i = 0; i < n; ++i) Arrays.fill(dp[i], Integer.MAX_VALUE);
        dp[src][0] = 0;
        for (int len = 1; len <= k + 1; ++len) {
            for (int i = 0; i < n; ++i) {
                if (dp[i][len - 1] != Integer.MAX_VALUE) {
                    for (Pair p : graph.get(i)) {
                        int v = p.getKey(), w = p.getValue();
                        dp[v][len] = Math.min(dp[v][len], dp[i][len - 1] + w);
                    }
                }
            }
        }
        int ans = Integer.MAX_VALUE;
        for (int len = 0; len < k + 2; ++len) ans = Math.min(ans, dp[dst][len]);
        return ans == Integer.MAX_VALUE ? -1 : ans;
    }
}